### ISSUE TITLE
PubSubGrainState JSON (de)serialization appears to have been broken

### ISSUE THREAD

**Issue by ca-ta** [dotnet/orleans#752]
Tested with 9c57a236eee9cf2a035300488cf0b12cb37c6138

To repro:
* configure `PubSubStore` to `Orleans.Storage.AzureTableStorage` and set `UseJsonFormat` to `true`.
* `PubSubRendesVousGrain` grains will appear to successfully save their state but will subsequently fail to deserialize it.

In the same test scenario, binary serialization works.

**Comment by gabikliot** [comment]
Thanks for reporting this Catalin!
Fixed via #754.

This Json serialization is so fragile! You change one field from public to private and it just skips serializing it, without any notice. 
And it is really hard to write an isolated unit test for it, since the data inside this state class is generated by the runtime and it's hard to fake it. We need to add a test with end to end scenario with json serialization.

**Comment by ca-ta** [comment]
@gabikliot - you're welcome - and thanks for the fix, I'll give it a whirl  sometime in the next couple of days.

I am too quite wary of the sneaky nature of the JSON serialization bugs - I haven't noticed this bug immediately, as most pub-sub scenarios continued to work. I luckily use very short grain lifetimes in testing.  I agree that BVT tests would be very helpful, but couldn't come up with one I was happy with in the limited time I was able to spend on this; that is why I filed without a repro - apologies for that.

I've added the option to switch between JSON and binary serialization (using the Orleans binary serializer) to my custom Azure Table storage provider, and that works. I'm not happy with loosing stored data legibility, but it will do for now.

I'm cool with closing this (or we can leave it open until the tests are in). Alas, I can't promise I'll have the bandwidth to add them in the next few weeks, it's crunch time around here :-). 

Thanks again!
